Add documentation for media browser image proxy (#699)
home-assistant/developers.home-assistant: docs/core/entity/media-player.md

| tv | Device is a television type device.
| speaker | Device is speakers or stereo type device.
| receiver | Device is audio video receiver type device taking audio and outputting to speakers and video to some display.

### Proxy album art for media browser

> Optional. Some media players can serve album artwork or other images on an internal network accessible to Home Assistant that is inaccessible from an external url.

Reviewer: ```suggestion
Optional. If your media player is only accessible from the internal network, it will need to proxy the album art via Home Assistant to be able to work while away from home.
```

> To proxy an image via Home Assistant, set the `thumbnail` property of a `BrowseMedia` item to a url generated by the `self.get_browse_image_url(media_content_type, media_content_id, media_image_id=None)` method. The browser will then fetch this url, which will result in a call to `async_get_browse_image(media_content_type, media_content_id, media_image_id=None)`.

Reviewer: ```suggestion
To proxy an image via Home Assistant, set the `thumbnail` property of a `BrowseMedia` item to a url generated by the `self.get_browse_image_url(media_content_type, media_content_id, media_image_id=None)` method. The browser will then fetch this url, which will result in a call to `async_get_browse_image(media_content_type, media_content_id, media_image_id=None)`.

> Only use a proxy for the thumbnail if the web request originated from outside the network. You can test this with `is_local_request(hass)` imported from `homeassistant.helpers.network`.
```

> To do so, implement the `async_get_browse_image` method. It takes the `media_content_type`, `media_content_id`, and `browse_image` `get` parameters. You may want to also create a method that generates the appropriate url and use it as part of your `async_browse_media` implementation. Note: `/api/media_player_proxy/{entity_id}` requires a `token` get parameter. The value of this token can be retrieved using the `access_token` property implemented by the `MediaPlayerEntity` base class.

Reviewer: ```suggestion
To proxy an image via Home Assistant, set the `thumbnail` property of a `BrowseMedia` item to a url generated by the `self.get_browse_image_url(media_content_type, media_content_id, media_image_id=None)` method. The browser will then fetch this url, which will result in a call to `async_get_browse_image(media_content_type, media_content_id, media_image_id=None)`.

In `async_get_browse_image`, use `self._async_fetch_image(url)` to fetch the image from the local network.

> Do not pass a url as `media_image_id`. This can allow an attacker to fetch any data from the local network.
```

```python
class MyMediaPlayer(MediaPlayerEntity):

>     # Implement the following method.

Reviewer: ```suggestion

    # Implement the following method.
```

>     async def async_get_browse_image(self, media_content_type, media_content_id, browse_image):

Reviewer: ```suggestion
    async def async_get_browse_image(self, media_content_type, media_content_id, media_image_id=None):
```

    """Serve album art. Returns (content, content_type)."""
    image_url = ...

>     result = await self_async_fetch_image(image_url)

Reviewer: ```suggestion
    result = await self._async_fetch_image(image_url)
```

    if result == (None, None):
        _LOGGER.info("Error retrieving proxied album art from %s", image_url)

>        return result

Reviewer: ```suggestion
   return result
```

>     result = await self._async_fetch_image(image_url)
>     if result == (None, None):
>         _LOGGER.info("Error retrieving proxied album art from %s", image_url)
>    return result

Reviewer: This check is now done by the media player proxy webview. 

```suggestion
    return await self._async_fetch_image(image_url)
```

> In your media browser implementation, use `self.get_browse_image_url(media_content_type, media_content_id, browse_image)` to generate the thumbnail paramater for `BrowseMedia`.

Reviewer: ```suggestion
In your media browser implementation, use `self.get_browse_image_url(media_content_type, media_content_id, media_image_id=None)` to generate the thumbnail paramater for `BrowseMedia`.
```
